### PR TITLE
Feature/export file menu option

### PR DESCRIPTION
**Pull request by manikag-12**
Implement export through the file menu option

Changes

- The export file menu option gets the filepath from the user to save the .edl file to
Screenshot (87)

### CODE REVIEW

File: src/main/handlers/exportProjectHandler.ts [hunk 1]
```diff
@@ -0,0 +1,48 @@
+import { writeFile } from 'fs/promises';
+import { BrowserWindow, dialog } from 'electron';
+import path from 'path';
+import { Project } from '../../sharedTypes';
+import { exportEDL } from '../export';
+
+const getSaveFilePath: (
```
Comment (chloebrett) [inline src/main/handlers/exportProjectHandler.ts]: This shouldn't be here

If it's the same as the one in the save handler, then just import that, if it's different then make your changes here and rename to getExportFilePath

File: src/main/handlers/exportProjectHandler.ts [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+  return dialogResponse.filePath as string;
+};
+
+const exportProjectToFile: (
+  filePath: string,
+  mainWindow: BrowserWindow | null,
+  project: Project
+) => Promise<void> = async (filePath, mainWindow, project) => {
+  // project.exportFilePath = path.dirname(filePath);
```
Comment (chloebrett) [inline src/main/handlers/exportProjectHandler.ts]: Remove

File: src/main/main.ts [hunk 3]
```diff
@@ -20,6 +20,7 @@ import startServer from './pyServer';
 import { appDataStoragePath, mkdir, resolveHtmlPath } from './util';
 import initialiseIpcHandlers from './ipc';
 import { extractAudio } from './handlers';
+// import handleExportProject from './exportProjectHandler';
```
Comment (chloebrett) [inline src/main/main.ts]: Remove

File: src/main/menu.ts [hunk 4]
```diff
@@ -157,6 +157,14 @@ export default class MenuBuilder {
         },
         enabled: false,
       },
+      {
+        id: 'export',
+        label: 'Export Project',
+        accelerator: 'CommandOrControl+Shift+E',
+        click: () => {
+          this.mainWindow.webContents.send('initiate-export-project');
```
Comment (chloebrett) [inline src/main/menu.ts:125]: Good 👍🏻  but this won't work on windows or linux, check the pattern used for the save / save as and undo / redo

File: src/main/handlers/exportProjectHandler.ts [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+  mainWindow: BrowserWindow | null
+) => Promise<string> = async (mainWindow) => {
+  if (mainWindow === null) {
+    throw new Error('Main window not defined');
+  }
+
+  const dialogResponse = await dialog.showSaveDialog(mainWindow, {
+    filters: [{ name: 'MLVET Files', extensions: ['mlvet'] }],
+    buttonLabel: 'Save',
+    title: 'Save Project',
```
Comment (chloebrett) [inline src/main/handlers/exportProjectHandler.ts]: This should be 'Export Project' etc

File: src/main/handlers/exportProjectHandler.ts [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+
+const getSaveFilePath: (
+  mainWindow: BrowserWindow | null
+) => Promise<string> = async (mainWindow) => {
+  if (mainWindow === null) {
+    throw new Error('Main window not defined');
+  }
+
+  const dialogResponse = await dialog.showSaveDialog(mainWindow, {
+    filters: [{ name: 'MLVET Files', extensions: ['mlvet'] }],
```
Comment (chloebrett) [inline src/main/handlers/exportProjectHandler.ts]: File type is edl not mlvet

File: src/renderer/ipc.ts [hunk 7]
```diff
@@ -67,3 +67,16 @@ window.electron.on('initiate-undo', async () => {
...
+ * Used by backend to initiate export from front end
+ */
+window.electron.on('initiate-export-project', async () => {
+  // Retrieve current project state from redux
+  const { currentProject } = store.getState();
+
+  // Don't export if we don't have a project open
+  if (currentProject === null) return;
+
+  await window.electron.exportProject(currentProject);
```
Comment (chloebrett) [inline src/renderer/ipc.ts]: Need to update the export file path on the front end as well, you can do this by returning the path used from the back end call

**Review by chloebrett — CHANGES_REQUESTED** [review]

File: src/main/handlers/exportProjectHandler.ts [hunk 8]
```diff
@@ -0,0 +1,48 @@
+import { writeFile } from 'fs/promises';
+import { BrowserWindow, dialog } from 'electron';
+import path from 'path';
```
Comment (exquisitebin) [inline src/main/handlers/exportProjectHandler.ts]: Remove unused imports

**Review by exquisitebin — COMMENTED** [review]
Patrick seems to have found everything that I can see. Once those changes are made I think it should be all good

**Comment by ghost** [comment]
Looks good. Will review once suggested changes have been resolved. 😃

File: src/main/menu.ts (hunk 4 above)
Comment (manikag-12) [reply src/main/menu.ts:125]: Tested on windows and it works. Pattern similar to save/save as

File: src/renderer/ipc.ts (hunk 7 above)
Comment (manikag-12) [reply src/renderer/ipc.ts]: Where on the front-end sorry?

Comment (ghost) [reply src/renderer/ipc.ts]: Have talked through this with @manikag-12

File: src/main/menu.ts (hunk 4 above)
Comment (chloebrett) [reply src/main/menu.ts:125]: Ah my bad, you're right this is piggy backing on the existing buildFileOptions()

File: src/renderer/ipc.ts (hunk 7 above)
Comment (chloebrett) [reply src/renderer/ipc.ts]: Okay cool. Basically end result is the redux store should be updated to include the export file path that the file was exported to.

File: src/main/menu.ts [hunk 9]
```diff
@@ -157,6 +157,14 @@ export default class MenuBuilder {
         },
         enabled: false,
       },
+      {
+        id: 'export',
+        label: 'Export Project',
+        accelerator: 'CommandOrControl+Shift+E',
```
Comment (chloebrett) [inline src/main/menu.ts]: I believe (unless this has changed) the key combo suggested was CommandOrControl+E for EDL export, and CommandOrControl+Shift+E for video export.

File: src/main/handlers/exportProjectHandler.ts [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+};
+
+const exportProjectToFile: (
+  filePath: string,
+  mainWindow: BrowserWindow | null,
+  project: Project
+) => Promise<void> = async (filePath, mainWindow, project) => {
+  project.exportFilePath = filePath;
+
+  exportEDL(mainWindow, project);
```
Comment (chloebrett) [inline src/main/handlers/exportProjectHandler.ts]: Would be really cool if we could make it open up the folder containing the export in finder/explorer after the export is done:

https://stackoverflow.com/questions/43991267/electron-open-file-directory-in-specific-application

File: src/main/handlers/exportProjectHandler.ts [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+import { exportEDL } from '../export';
+
+const getExportFilePath: (
+  mainWindow: BrowserWindow | null
+) => Promise<string> = async (mainWindow) => {
+  if (mainWindow === null) {
+    throw new Error('Main window not defined');
+  }
+
+  const dialogResponse = await dialog.showSaveDialog(mainWindow, {
```
Comment (chloebrett) [inline src/main/handlers/exportProjectHandler.ts]: Give it a sensible default file name e.g. {projectName}.edl

**Review by chloebrett — CHANGES_REQUESTED** [review]
Getting close, just a couple of things

File: src/main/menu.ts (hunk 9 above)
Comment (manikag-12) [reply src/main/menu.ts]: Not sure where this was mentioned

File: src/main/handlers/exportProjectHandler.ts (hunk 10 above)
Comment (manikag-12) [reply src/main/handlers/exportProjectHandler.ts]: Seems pretty cool but is out of scope of this feature for MVP. Will create a ticket in the backlog after checking with other PMs

File: src/main/menu.ts (hunk 9 above)
Comment (manikag-12) [reply src/main/menu.ts]: My bad just checked figma. Updated :))

File: src/renderer/store/exportIo/actions.ts [hunk 12]
```diff
@@ -4,9 +4,15 @@ export const START_EXPORT = 'START_EXPORT';
 export const EXPORT_PROGRESS_UPDATE = 'EXPORT_PROGRESS_UPDATE';
 export const FINISH_EXPORT = 'FINISH_UPDATE';
 
-export const startExport: () => Action<null> = () => ({
+export const startExport: (
```
Comment (ghost) [inline src/renderer/store/exportIo/actions.ts:7]: Looks like the redux isn't getting updated. Is this action getting dispatched at all?

File: src/main/handlers/exportProjectHandler.ts [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+  });
+
+  if (dialogResponse.canceled) {
+    throw new Error('Dialog cancelled');
+  }
+
+  return dialogResponse.filePath as string;
+};
+
+const exportProjectToFile: (
```
Comment (ghost) [inline src/main/handlers/exportProjectHandler.ts]: I think the type syntax for the functions in this file should be the same as for the other handlers. There was a PR to make them all consistent in preparation for the IPC handler gen code PR.

File: src/main/ipc.ts [hunk 14]
```diff
@@ -81,6 +82,10 @@ const initialiseIpcHandlers: (ipcContext: IpcContext) => void = (
     writeRecentProjects(recentProjects)
   );
   // END GENERATED CODE PART 2
+  ipcMain.handle('user-os', async () => handleOsQuery());
```
Comment (ghost) [inline src/main/ipc.ts]: Is this line a duplicate from a merge conflict?

File: src/main/ipc.ts [hunk 15]
```diff
@@ -81,6 +82,10 @@ const initialiseIpcHandlers: (ipcContext: IpcContext) => void = (
...
+  ipcMain.handle('export-project', async (_event, project) =>
```
Comment (ghost) [inline src/main/ipc.ts:85]: I believe this handler should go inside the generated code block

File: src/renderer/store/exportIo/actions.ts (hunk 12 above)
Comment (manikag-12) [reply src/renderer/store/exportIo/actions.ts:7]: ah fk I think I messed up while rebasing. Fixing now

File: src/renderer/ipcRegistration.ts [hunk 16]
```diff
@@ -58,3 +58,16 @@ ipc.on('initiate-undo', async () => {
...
+/**
+ * Used by backend to initiate export from front end
+ */
+ipc.on('initiate-export-project', async () => {
+  // Retrieve current project state from redux
+  const { currentProject } = store.getState();
+
+  // Don't export if we don't have a project open
+  if (currentProject === null) return;
+
```
Comment (ghost) [inline src/renderer/ipcRegistration.ts:75]: Have a look at 'initiate-save-project'. I think export needs to be split into 2 separate functions. 1 - to get the save file path from the backend (then dispatch that to redux). 2 - to export the project using the backend.

File: src/renderer/store/exportIo/actions.ts (hunk 12 above)
Comment (manikag-12) [reply src/renderer/store/exportIo/actions.ts:7]: done. getting dispatched from "initiate-export-project" [here](https://github.com/patrickbrett/mlvet/pull/56/files#:~:text=store.dispatch(startExport(currentProject.id%2C%20filePath))%3B)

File: src/renderer/ipcRegistration.ts (hunk 16 above)
Comment (manikag-12) [reply src/renderer/ipcRegistration.ts:75]: I think the logic is identical here for both atm. Unless I missed something?

File: src/main/handlers/exportProjectHandler.ts (hunk 13 above)
Comment (manikag-12) [reply src/main/handlers/exportProjectHandler.ts]: done

File: src/main/ipc.ts (hunk 14 above)
Comment (manikag-12) [reply src/main/ipc.ts]: yep. removed

File: src/renderer/ipcRegistration.ts (hunk 16 above)
Comment (ghost) [reply src/renderer/ipcRegistration.ts:75]: image

Comment (manikag-12) [reply src/renderer/ipcRegistration.ts:75]: wait so exportEDL should be called here? and not in [exportProjectToFile](https://github.com/patrickbrett/mlvet/pull/56/files#:~:text=export%20const%20exportProjectToFile%3A%20()

**Review by chloebrett — APPROVED** [review]
LGTM, and nice work using the new handler style. Apologies for requiring like four refactors due to codebase architecture changes in the time this has been open 😅

**Comment by manikag-12** [comment]
okay refactor issues finally resolved!! Please let me know if this is good to go because I cannot refactor again 😭😭😭😭

**Review by ghost — APPROVED** [review]

**Pull request merged**